text/0567-use-and-resources.md

```diff
@@ -0,0 +1,953 @@
+  `update` will trigger another `update`.
+
+#### `teardown`
+
+A lifecycle hook that tears the `Resource` instance down.
+
+- If no `update` method is defined, runs on the previous instance of
+  `Resource` whenever args or state tracked in `start` updates, before a new
+  instance of the `Resource` is created.
+- Runs before the parent that the `Resource` belongs to is destroyed.
+
+#### TypeScript Support
+
+TypeScript is not officially supported by Ember, but as its adoption grows it's
+a good idea to make sure that the APIs we are designing _can_ be typed,
+especially with APIs that use decorators. TypeScript users have historically had
+many issues with APIs that use decorators and are similar to `@use`, since
+decorators cannot change types currently. TS support for
+`ember-concurrency-decorators` for instance, is very difficult.
+
+Since `@use` takes an instance of a class that is assigned to a class field and
+instead makes the field return the `state` property of that class, it could face
+similar issues here:
+
+```ts
+class MyState extends Resource {
+  state: number = 123;
+}
+
+class MyComponent extends Component {
+  // This field will be a readonly number, but if `create` is being accurate,
+  // TS will _think_ it is an instance of MyState
+  @use myState = MyState.create();
+}
+```
+
+However, we can fudge the types a bit here to make it work better for TS:
+
+```ts
+declare class Resource<T> {
+  state: T;
+
+  static create(): T;
+}
+```
+
+From most user's perspectives, the return value of the static `create` method
+should be opaque anyways, and doing this will allow TS to "just work" with
+TypeScript out of the box.
+
+### Module API
+
+The APIs described in this RFC will be importable from the `@glimmer/tracking`
+package:
+
+```js
+import { use } from '@glimmer/tracking';
+import Resource, {
+  setResourceManager,
+  getUseInfo,
+  setUseArgs,
+  scheduleUseStart,
+} from '@glimmer/tracking/resource';
+```
+
+### Name Choice
+
+The term "resource" is used in many programming languages to refer to a value
+which inherently has a managed lifecycle. Some examples include
+
+* The [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization)
+  pattern, which the design of resources was heavily influenced by (with the
+  addition of an "update").
+* The [try-with-resource](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)
+  syntax from Java, and the [explicit-resource-management](https://github.com/tc39/proposal-explicit-resource-management)
+  proposal for JavaScript.
+
+## How we teach this
+
+`@use` and resources are a new concept in Ember Octane, but they are
+fundamentally related to the existing state flow patterns that were established.
+They should be introduced in the `Components` section of the guides, after
+modifiers. The docs should include useful examples for how to accomplish basic
+tasks, such as writing a `RemoteData` resource.
```

Comment:
As this RFC not only proposes the low-level manager API, but also a ready to use `Resource` class, what does this imply for Ember CLI?
* will `ember g resource` and `ember g resource-test` be introduced?
* will there be an "official" new app folder `/resources` where resource classes live by default?

Reply:
That's a great question. I think generating the resource and resource test would be good, though I'm not sure where they should go.

I don't think we want to add a new top level `/resources` folder - my aim was to not add resources as a top-level resolvable, and to encourage defining them inline or nearby to the place they are used, more like utilities. There will definitely be some common resources that are used in many places, but there will also likely be some resources that are one-offs, only used in one place in the app.

Maybe they could go under `/app/utils/resources`? This really starts to get into how generators will work in general in a more import-driven world.

Reply:
Disagree with putting this in a Components section of the guides. This seems complex enough and useful enough to deserve its own section in the guides. IMO this is a top level construct similar to a React Context.

Reply:
> That's a great question. I think generating the resource and resource test would be good, though I'm not sure where they should go.
> 
> I don't think we want to add a new top level `/resources` folder - my aim was to not add resources as a top-level resolvable, and to encourage defining them inline or nearby to the place they are used, more like utilities. There will definitely be some common resources that are used in many places, but there will also likely be some resources that are one-offs, only used in one place in the app.
> 
> Maybe they could go under `/app/utils/resources`? This really starts to get into how generators will work in general in a more import-driven world.

I agree resources should not be resolvable by the resolver. They should be imported as described here. However, I think for organizational purposes it should be _recommended_ that a top level resources folder be created and used for resources.

IMO this is a top level construct similar to a React Context.